File: app/lib/api/s3.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../logger.dart';
import '../services/s3_local_test.dart';
import 'client.dart';

enum S3StorageMode {
  /// 当前部署没有内置 S3，且该用户也未配置自建 S3。
  disabled,

  /// 该用户走平台内置 S3（仅海外集群且 hosted 桶启用时可见）。
  hosted,

  /// 该用户已经配置了自建 S3。
  custom;

  static S3StorageMode parse(dynamic raw, {required bool fallbackConfigured}) {
    if (raw is String) {
      switch (raw.toUpperCase()) {
        case 'CUSTOM':
          return S3StorageMode.custom;
        case 'HOSTED':
          return S3StorageMode.hosted;
        case 'DISABLED':
          return S3StorageMode.disabled;
      }
    }
    return fallbackConfigured ? S3StorageMode.custom : S3StorageMode.disabled;
  }
}

class S3ConfigDetail {
  final bool configured;
  final S3StorageMode mode;
  final bool hostedAvailable;
  final bool customSaved;
  final String? endpoint;
  final String? region;
  final String? bucket;
  final String? accessKeyId;
  final bool? pathStyleAccessEnabled;
  final String? documentationUrl;

  S3ConfigDetail({
    required this.configured,
    required this.mode,
    required this.hostedAvailable,
    this.customSaved = false,
    this.endpoint,
    this.region,
    this.bucket,
    this.accessKeyId,
    this.pathStyleAccessEnabled,
    this.documentationUrl,
  });

  factory S3ConfigDetail.fromJson(Map<String, dynamic> j) {
    final fallbackConfigured = (j['configured'] as bool?) ?? false;
    final mode = S3StorageMode.parse(
      j['mode'],
      fallbackConfigured: fallbackConfigured,
    );
    final customSaved = (j['customSaved'] as bool?) ?? (mode == S3StorageMode.custom);
    return S3ConfigDetail(
      configured: mode != S3StorageMode.disabled,
      mode: mode,
      hostedAvailable: (j['hostedAvailable'] as bool?) ?? false,
      customSaved: customSaved,
      endpoint: j['endpoint'] as String?,
      region: j['region'] as String?,
      bucket: j['bucket'] as String?,
      accessKeyId: j['accessKeyId'] as String?,
      pathStyleAccessEnabled: j['pathStyleAccessEnabled'] as bool?,
      documentationUrl: j['documentationUrl'] as String?,
    );
  }

  factory S3ConfigDetail.disabled() => S3ConfigDetail(
    configured: false,
    mode: S3StorageMode.disabled,
    hostedAvailable: false,
  );
}

Future<S3ConfigDetail> getS3Config() async {
  await clearLegacyS3ConfigCache();
  if (!hasAccessToken) {
    logApi.fine('getS3Config no token');
    return S3ConfigDetail.disabled();
  }
  return withAuthRetry(() async {
    final r = await http.get(
      Uri.parse('$apiBaseUrl/api/s3/config'),
      headers: apiHeaders,
    );
    if (r.statusCode == 401) throw AuthException();
    if (r.statusCode != 200) return S3ConfigDetail.disabled();
    final detail = S3ConfigDetail.fromJson(
      jsonDecode(r.body) as Map<String, dynamic>,
    );
    logApi.info(
      'getS3Config mode=${detail.mode.name} hostedAvailable=${detail.hostedAvailable}',
    );
    return detail;
  });
}

Future<bool> hasS3Config() async {
  final detail = await getS3Config();
  return detail.configured;
}

class S3ConfigRequest {
  final String endpoint;
  final String? region;
  final String bucket;
  final String accessKeyId;
  final String secretAccessKey;
  final bool pathStyleAccessEnabled;

  S3ConfigRequest({
    required this.endpoint,
    this.region,
    required this.bucket,
    required this.accessKeyId,
    required this.secretAccessKey,
    this.pathStyleAccessEnabled = true,
  });

  Map<String, dynamic> toJson() => {
    'endpoint': endpoint,
    if (region != null && region!.isNotEmpty) 'region': region,
    'bucket': bucket,
    'accessKeyId': accessKeyId,
    if (secretAccessKey.isNotEmpty) 'secretAccessKey': secretAccessKey,
    'pathStyleAccessEnabled': pathStyleAccessEnabled,
  };
}

Future<void> saveS3Config(S3ConfigRequest req) async {
  logApi.info('saveS3Config endpoint=${req.endpoint} bucket=${req.bucket}');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/config'),
      headers: apiHeaders,
      body: jsonEncode(req.toJson()),
    );
    checkAuthResponse(r, fallback: '保存 S3 配置失败');
    if (r.statusCode != 204) throw Exception('保存失败');
    logApi.info('saveS3Config success');
  });
}

Future<void> testS3Config() async {
  logApi.info('testS3Config');
  final detail = await getS3Config();
  if (detail.mode == S3StorageMode.hosted) return;
  if (detail.mode != S3StorageMode.custom) {
    throw Exception('S3 连接测试失败');
  }

  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/test'),
      headers: apiHeaders,
      body: '{}',
    );
    checkAuthResponse(r, fallback: 'S3 连接测试失败');
    if (r.statusCode != 200) {
      try {
        final body = jsonDecode(r.body) as Map<String, dynamic>;
        final err = body['error'] as String?;
        if (err != null && err.isNotEmpty) throw Exception(err);
      } catch (e) {
        if (e is Exception) rethrow;
      }
      throw Exception('S3 连接测试失败');
    }
    final data = jsonDecode(r.body) as Map<String, dynamic>;
    final url = data['url'] as String?;
    if (url == null || url.isEmpty) throw Exception('S3 连接测试失败');
    await headPresignedUrl(url);
    logApi.info('testS3Config success');
  });
}

Future<bool> checkS3Online() async {
  final detail = await getS3Config();
  if (!detail.configured) return false;
  if (detail.mode == S3StorageMode.hosted) return true;
  if (detail.mode != S3StorageMode.custom) return false;
  try {
    await testS3Config();
    return true;
  } catch (_) {
    return false;
  }
}

Future<void> clearS3Config() async {
  logApi.info('clearS3Config');
  return withAuthRetry(() async {
    final r = await http.delete(
      Uri.parse('$apiBaseUrl/api/s3/config'),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '清空 S3 配置失败');
    logApi.info('clearS3Config success');
  });
}

Future<void> useHostedS3() async {
  logApi.info('useHostedS3');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/use-hosted'),
      headers: apiHeaders,
      body: '{}',
    );
    checkAuthResponse(r, fallback: '切换到内置 S3 失败');
    logApi.info('useHostedS3 success');
  });
}

Future<void> useCustomS3() async {
  logApi.info('useCustomS3');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/use-custom'),
      headers: apiHeaders,
      body: '{}',
    );
    checkAuthResponse(r, fallback: '切换到自建 S3 失败');
    logApi.info('useCustomS3 success');
  });
}

class PresignUploadResponse {
  final String uploadUrl;
  final String key;

  PresignUploadResponse({required this.uploadUrl, required this.key});

  factory PresignUploadResponse.fromJson(Map<String, dynamic> j) =>
      PresignUploadResponse(
        uploadUrl: j['uploadUrl'] as String,
        key: j['key'] as String,
      );
}

Future<PresignUploadResponse> presignUpload(
  String fileName, {
  String? contentType,
  int? contentLength,
}) async {
  logApi.info('presignUpload fileName=$fileName');
  return withAuthRetry(() async {
    final body = <String, dynamic>{'fileName': fileName};
    if (contentType != null && contentType.isNotEmpty) {
      body['contentType'] = contentType;
    }
    if (contentLength != null && contentLength > 0) {
      body['contentLength'] = contentLength;
    }
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/presign-upload'),
      headers: apiHeaders,
      body: jsonEncode(body),
    );
    checkAuthResponse(r, fallback: '获取上传地址失败');
    final res = PresignUploadResponse.fromJson(
      jsonDecode(r.body) as Map<String, dynamic>,
    );
    logApi.info('presignUpload success key=${res.key}');
    return res;
  });
}

Future<String> getDownloadUrl(String key) async {
  logApi.info('getDownloadUrl key=$key');
  return withAuthRetry(() async {
    final r = await http.get(
      Uri.parse('$apiBaseUrl/api/s3/download-url').replace(
        queryParameters: {'key': key},
      ),
      headers: apiHeaders,
    );
    checkAuthResponse(r, fallback: '获取下载地址失败');
    final data = jsonDecode(r.body) as Map<String, dynamic>;
    return data['url'] as String;
  });
}

class MultipartInitiateResponse {
  final String uploadId;
  final String key;

  MultipartInitiateResponse({required this.uploadId, required this.key});

  factory MultipartInitiateResponse.fromJson(Map<String, dynamic> j) =>
      MultipartInitiateResponse(
        uploadId: j['uploadId'] as String,
        key: j['key'] as String,
      );
}

Future<MultipartInitiateResponse> initiateMultipartUpload(
  String fileName, {
  String? contentType,
  int? totalSize,
}) async {
  logApi.info('initiateMultipart fileName=$fileName');
  return withAuthRetry(() async {
    final body = <String, dynamic>{'fileName': fileName};
    if (contentType != null && contentType.isNotEmpty) {
      body['contentType'] = contentType;
    }
    if (totalSize != null && totalSize > 0) body['totalSize'] = totalSize;
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/multipart/initiate'),
      headers: apiHeaders,
      body: jsonEncode(body),
    );
    checkAuthResponse(r, fallback: '创建分片上传失败');
    return MultipartInitiateResponse.fromJson(
      jsonDecode(r.body) as Map<String, dynamic>,
    );
  });
}

Future<String> presignUploadPart({
  required String uploadId,
  required String key,
  required int partNumber,
}) async {
  logApi.fine('presignPart part=$partNumber');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/multipart/presign-part'),
      headers: apiHeaders,
      body: jsonEncode({
        'uploadId': uploadId,
        'key': key,
        'partNumber': partNumber,
      }),
    );
    checkAuthResponse(r, fallback: '获取分片上传地址失败');
    final data = jsonDecode(r.body) as Map<String, dynamic>;
    return data['url'] as String;
  });
}

Future<void> completeMultipartUpload({
  required String uploadId,
  required String key,
  required List<Map<String, dynamic>> parts,
}) async {
  logApi.info('completeMultipart key=$key parts=${parts.length}');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/multipart/complete'),
      headers: apiHeaders,
      body: jsonEncode({'uploadId': uploadId, 'key': key, 'parts': parts}),
    );
    checkAuthResponse(r, fallback: '完成分片上传失败');
  });
}

Future<void> abortMultipartUpload({
  required String uploadId,
  required String key,
}) async {
  logApi.info('abortMultipart key=$key');
  return withAuthRetry(() async {
    final r = await http.post(
      Uri.parse('$apiBaseUrl/api/s3/multipart/abort'),
      headers: apiHeaders,
      body: jsonEncode({'uploadId': uploadId, 'key': key}),
    );
    checkAuthResponse(r, fallback: '取消分片上传失败');
  });
}

File: web/src/lib/api/s3.ts
import { logger } from '../logger';
import { getApiUrl, TAG, AuthError, getToken, isAuthFailure, withAuthRetry } from './client';
import { clearLegacyS3ConfigCache, headPresignedUrl } from '../services/s3LocalTest';

export type S3ConfigRequest = {
  endpoint: string;
  region?: string;
  bucket: string;
  accessKeyId: string;
  /** 留空则保存时不修改已有 secret */
  secretAccessKey?: string;
  /** true = path-style; false = virtual-hosted. Defaults to true when omitted. */
  pathStyleAccessEnabled?: boolean;
};

/**
 * 当前用户的 S3 存储模式。
 * - `DISABLED`：当前部署没有内置 S3，且该用户也未配置自建 S3。
 * - `HOSTED`：该用户走平台内置 S3（仅海外集群且 hosted 桶启用时可见）。
 * - `CUSTOM`：该用户已经配置了自建 S3。
 */
export type S3StorageMode = 'DISABLED' | 'HOSTED' | 'CUSTOM';

export type S3ConfigResponse = {
  /** Convenience flag, equivalent to `mode != 'DISABLED'`. */
  configured: boolean;
  mode: S3StorageMode;
  /** 当前部署是否提供了内置 S3（决定是否展示「切换回内置 S3」按钮）。 */
  hostedAvailable: boolean;
  /**
   * 用户后端是否保存过自建 S3 凭证。HOSTED 模式下若为 true，表示曾经配置过、
   * 当前主动选择走内置 S3，可一键「使用已保存的自建 S3」切回，无需重新输入。
   */
  customSaved: boolean;
  /** 当前部署下 S3 配置说明文档绝对 URL（Flutter 外开）。 */
  documentationUrl?: string;
  endpoint?: string;
  region?: string;
  bucket?: string;
  accessKeyId?: string;
  pathStyleAccessEnabled?: boolean;
};

function disabledResponse(): S3ConfigResponse {
  return { configured: false, mode: 'DISABLED', hostedAvailable: false, customSaved: false };
}

function normalizeS3Response(raw: unknown): S3ConfigResponse {
  const obj = (raw ?? {}) as Record<string, unknown>;
  const fallbackConfigured = obj.configured === true;
  const rawMode = typeof obj.mode === 'string' ? obj.mode.toUpperCase() : null;
  const mode: S3StorageMode =
    rawMode === 'CUSTOM' || rawMode === 'HOSTED' || rawMode === 'DISABLED'
      ? (rawMode as S3StorageMode)
      : fallbackConfigured
        ? 'CUSTOM'
        : 'DISABLED';
  // 兼容旧后端：CUSTOM 模式必然意味着已保存
  const customSaved = obj.customSaved === true || mode === 'CUSTOM';
  return {
    configured: mode !== 'DISABLED',
    mode,
    hostedAvailable: obj.hostedAvailable === true,
    customSaved,
    documentationUrl: typeof obj.documentationUrl === 'string' ? obj.documentationUrl : undefined,
    endpoint: typeof obj.endpoint === 'string' ? obj.endpoint : undefined,
    region: typeof obj.region === 'string' ? obj.region : undefined,
    bucket: typeof obj.bucket === 'string' ? obj.bucket : undefined,
    accessKeyId: typeof obj.accessKeyId === 'string' ? obj.accessKeyId : undefined,
    pathStyleAccessEnabled:
      typeof obj.pathStyleAccessEnabled === 'boolean' ? obj.pathStyleAccessEnabled : undefined,
  };
}

export async function getS3Config(): Promise<S3ConfigResponse> {
  clearLegacyS3ConfigCache();
  if (!getToken()) {
    logger.debug(TAG, 'getS3Config: no token');
    return disabledResponse();
  }
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) return disabledResponse();
    const res = await fetch(`${getApiUrl()}/api/s3/config`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) return disabledResponse();
    const data = normalizeS3Response(await res.json());
    logger.info(TAG, 'getS3Config mode=', data.mode, 'hostedAvailable=', data.hostedAvailable);
    return data;
  });
}

export async function hasS3Config(): Promise<boolean> {
  const data = await getS3Config();
  return data.configured;
}

export async function saveS3Config(data: S3ConfigRequest): Promise<void> {
  logger.info(TAG, 'saveS3Config endpoint=', data.endpoint, 'bucket=', data.bucket);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/config`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(data),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'saveS3Config failed', res.status);
      throw new Error('Failed to save S3 config');
    }
    logger.info(TAG, 'saveS3Config success');
  });
}

export async function clearS3Config(): Promise<void> {
  logger.info(TAG, 'clearS3Config');
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/config`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'clearS3Config failed', res.status);
      throw new Error('Failed to clear S3 config');
    }
    logger.info(TAG, 'clearS3Config success');
  });
}

/** 切换到平台内置 S3，但保留已保存的自建 S3 凭证。 */
export async function switchToHostedS3(): Promise<void> {
  logger.info(TAG, 'switchToHostedS3');
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/use-hosted`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: '{}',
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'switchToHostedS3 failed', res.status);
      throw new Error('Failed to switch to hosted S3');
    }
    logger.info(TAG, 'switchToHostedS3 success');
  });
}

/** 切换到此前保存的自建 S3 配置（要求后端已存在 BYO 凭证）。 */
export async function switchToCustomS3(): Promise<void> {
  logger.info(TAG, 'switchToCustomS3');
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/use-custom`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: '{}',
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'switchToCustomS3 failed', res.status);
      throw new Error('Failed to switch to custom S3');
    }
    logger.info(TAG, 'switchToCustomS3 success');
  });
}

async function fetchS3TestUrl(): Promise<string> {
  const token = getToken();
  if (!token) throw new Error('errors.notAuthenticated');
  const res = await fetch(`${getApiUrl()}/api/s3/test`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
    body: '{}',
  });
  if (isAuthFailure(res)) throw new AuthError();
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    const msg = (err as { error?: string }).error || 'errors.s3TestFailed';
    throw new Error(msg);
  }
  const data = (await res.json()) as { url?: string };
  if (!data.url) throw new Error('errors.s3TestFailed');
  return data.url;
}

/** CUSTOM：服务端签发 HeadBucket 预签名 URL，本机 HEAD 探测。HOSTED：无需测试。 */
export async function testS3Config(): Promise<void> {
  logger.info(TAG, 'testS3Config');
  const cfg = await getS3Config();
  if (cfg.mode === 'HOSTED') return;
  if (cfg.mode !== 'CUSTOM') throw new Error('errors.s3TestFailed');

  return withAuthRetry(async () => {
    const url = await fetchS3TestUrl();
    await headPresignedUrl(url);
    logger.info(TAG, 'testS3Config success');
  });
}

/** 用于在线状态：HOSTED 已配置即在线；CUSTOM 需本机 HEAD 通过。 */
export async function checkS3Online(): Promise<boolean> {
  const cfg = await getS3Config();
  if (!cfg.configured) return false;
  if (cfg.mode === 'HOSTED') return true;
  if (cfg.mode !== 'CUSTOM') return false;
  try {
    await testS3Config();
    return true;
  } catch {
    return false;
  }
}

export type PresignUploadResponse = { uploadUrl: string; key: string };

export async function presignUpload(fileName: string, contentType?: string, contentLength?: number): Promise<PresignUploadResponse> {
  logger.info(TAG, 'presignUpload fileName=', fileName);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const body: Record<string, unknown> = { fileName, contentType };
    if (contentLength != null && contentLength > 0) body.contentLength = contentLength;
    const res = await fetch(`${getApiUrl()}/api/s3/presign-upload`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'presignUpload failed', res.status);
      throw new Error('Failed to get upload URL');
    }
    const data = await res.json() as PresignUploadResponse;
    logger.info(TAG, 'presignUpload success key=', data.key);
    return data;
  });
}

export async function getDownloadUrl(key: string): Promise<string> {
  logger.info(TAG, 'getDownloadUrl key=', key);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/download-url?key=${encodeURIComponent(key)}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) {
      logger.warn(TAG, 'getDownloadUrl failed', res.status);
      throw new Error('Failed to get download URL');
    }
    const data = (await res.json()) as { url: string };
    return data.url;
  });
}

// ── Multipart Upload ───────────────────────────────────────────────

export type MultipartInitiateResponse = { uploadId: string; key: string };

export async function initiateMultipartUpload(
  fileName: string,
  contentType?: string,
  totalSize?: number,
): Promise<MultipartInitiateResponse> {
  logger.info(TAG, 'initiateMultipart fileName=', fileName);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const body: Record<string, unknown> = { fileName, contentType };
    if (totalSize != null && totalSize > 0) body.totalSize = totalSize;
    const res = await fetch(`${getApiUrl()}/api/s3/multipart/initiate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify(body),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) throw new Error('Failed to initiate multipart upload');
    return (await res.json()) as MultipartInitiateResponse;
  });
}

export async function presignUploadPart(
  uploadId: string,
  key: string,
  partNumber: number,
): Promise<string> {
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/multipart/presign-part`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify({ uploadId, key, partNumber }),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) throw new Error('Failed to presign part');
    const data = (await res.json()) as { url: string };
    return data.url;
  });
}

export async function completeMultipartUpload(
  uploadId: string,
  key: string,
  parts: Array<{ partNumber: number; eTag: string }>,
): Promise<void> {
  logger.info(TAG, 'completeMultipart key=', key, 'parts=', parts.length);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/multipart/complete`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify({ uploadId, key, parts }),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) throw new Error('Failed to complete multipart upload');
  });
}

export async function abortMultipartUpload(
  uploadId: string,
  key: string,
): Promise<void> {
  logger.info(TAG, 'abortMultipart key=', key);
  return withAuthRetry(async () => {
    const token = getToken();
    if (!token) throw new Error('errors.notAuthenticated');
    const res = await fetch(`${getApiUrl()}/api/s3/multipart/abort`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${token}` },
      body: JSON.stringify({ uploadId, key }),
    });
    if (isAuthFailure(res)) throw new AuthError();
    if (!res.ok) throw new Error('Failed to abort multipart upload');
  });
}
